File: api/fill-precios.py
# api/fill-precios.py
# Vercel Serverless Function (Python runtime)
# Lee el archivo de cambios de precios, rellena la plantilla elegida y la devuelve

from http.server import BaseHTTPRequestHandler
import json, base64, io, cgi, os

def fill_template(template_bytes, template_name, products):
    """
    Rellena la plantilla reemplazando DESCRIPCION y PRECIO con los datos reales.
    Conserva logos, imágenes, formato y estructura intactos.
    """
    from openpyxl import load_workbook
    from openpyxl.worksheet.worksheet import Worksheet

    wb = load_workbook(
        io.BytesIO(template_bytes),
        read_only=False,
        data_only=False,
        keep_vba='xlsm' in template_name.lower()
    )

    prod_idx = 0  # índice del siguiente producto a colocar

    for sheet_name in wb.sheetnames:
        ws = wb[sheet_name]
        if not isinstance(ws, Worksheet):
            continue

        # Encontrar todas las celdas DESCRIPCION y PRECIO en orden
        desc_cells  = []
        price_cells = []

        for r in range(1, ws.max_row + 1):
            for c in range(1, ws.max_column + 1):
                cell = ws.cell(row=r, column=c)
                v = cell.value
                if v is None:
                    continue
                vs = str(v).strip().upper()
                if vs == 'DESCRIPCION':
                    desc_cells.append((r, c))
                elif vs == 'PRECIO':
                    price_cells.append((r, c))

        # Cada par (desc_cell, price_cell) es un slot
        slots = list(zip(desc_cells, price_cells))

        for (dr, dc), (pr, pc) in slots:
            if prod_idx >= len(products):
                # Sin más productos — limpiar la celda
                ws.cell(row=dr, column=dc).value = ''
                ws.cell(row=pr, column=pc).value = ''
            else:
                p = products[prod_idx]
                # Descripción
                ws.cell(row=dr, column=dc).value = p['name']
                # Precio en formato moneda: $489.00
                price_val = p['price']
                ws.cell(row=pr, column=pc).value = price_val
                # Formato de número moneda con 2 decimales
                ws.cell(row=pr, column=pc).number_format = '"$"#,##0.00'
                prod_idx += 1

    out = io.BytesIO()
    wb.save(out)
    return out.getvalue()


def parse_cambios(file_bytes):
    """
    Lee el archivo de cambios de precios.
    Última hoja, col B = descripción, col D = precio, desde fila 3.
    """
    from openpyxl import load_workbook

    wb = load_workbook(io.BytesIO(file_bytes), read_only=True, data_only=True)
    sheet_name = wb.sheetnames[-1]  # última hoja
    ws = wb[sheet_name]

    products = []
    for r in range(3, (ws.max_row or 200) + 1):
        name  = ws.cell(row=r, column=2).value  # col B
        price = ws.cell(row=r, column=4).value  # col D
        if name is None or str(name).strip() == '':
            continue
        name  = str(name).strip().upper()
        price = float(str(price).replace(',', '.').replace('$', '').strip()) if price is not None else None
        if price is None:
            continue
        products.append({'name': name, 'price': price})

    return products, sheet_name


class handler(BaseHTTPRequestHandler):

    def do_OPTIONS(self):
        self.send_response(200)
        self._cors()
        self.end_headers()

    def do_POST(self):
        self._cors()
        content_type = self.headers.get('Content-Type', '')

        try:
            # Leer body JSON con los archivos en base64
            length = int(self.headers.get('Content-Length', 0))
            body   = json.loads(self.rfile.read(length))

            cambios_b64  = body['cambios']     # base64 del archivo de cambios
            template_key = body['template']    # 'iman' o 'sin_iman'

            # Decodificar archivo de cambios
            cambios_bytes = base64.b64decode(cambios_b64)

            # Leer productos del archivo de cambios
            products, sheet_name = parse_cambios(cambios_bytes)

            if not products:
                self._json(400, {'error': f'No se encontraron productos en la hoja "{sheet_name}"'})
                return

            # Cargar plantilla desde carpeta /plantillas
            base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
            templates = {
                'iman':     os.path.join(base_dir, 'plantillas', 'PRECIOS_IMAN.xlsm'),
                'sin_iman': os.path.join(base_dir, 'plantillas', 'PRECIOS_SIN_IMAN.xlsx'),
            }

            if template_key not in templates:
                self._json(400, {'error': f'Plantilla desconocida: {template_key}'})
                return

            template_path = templates[template_key]
            with open(template_path, 'rb') as f:
                template_bytes = f.read()

            template_name = os.path.basename(template_path)

            # Rellenar plantilla
            result_bytes = fill_template(template_bytes, template_name, products)

            # Responder con el archivo rellenado en base64
            result_b64 = base64.b64encode(result_bytes).decode()
            ext = 'xlsm' if template_key == 'iman' else 'xlsx'
            filename = f'PRECIOS_{template_key.upper()}_{sheet_name.replace(" ","-")}.{ext}'

            self._json(200, {
                'ok': True,
                'file': result_b64,
                'filename': filename,
                'count': len(products),
                'sheet': sheet_name,
                'ext': ext,
            })

        except Exception as e:
            import traceback
            self._json(500, {'error': str(e), 'trace': traceback.format_exc()})

    def _cors(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')

    def _json(self, status, data):
        body = json.dumps(data).encode()
        self.send_response(status)
        self._cors()
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        self.wfile.write(body)

File: api/fill-precios.js
// api/fill-precios.js
// Edita el ZIP del Excel directamente — cambia solo las celdas DESCRIPCION y PRECIO.
// Logos, imágenes, estilos y macros quedan byte a byte idénticos al original.

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import AdmZip from 'adm-zip';
import * as XLSX from 'xlsx';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// Índices de shared strings en cada plantilla
// IMAN:     idx4=DESCRIPCION, idx5=PRECIO
// SIN_IMAN: idx2=DESCRIPCION, idx3=PRECIO
const TPL_CONFIG = {
  iman: {
    file:      'PRECIOS_IMAN.xlsm',
    sheets:    ['xl/worksheets/sheet1.xml'],
    idxDesc:   4,
    idxPrecio: 5,
    ext:       'xlsm',
  },
  sin_iman: {
    file:      'PRECIOS_SIN_IMAN.xlsx',
    sheets:    ['xl/worksheets/sheet1.xml'],
    idxDesc:   2,
    idxPrecio: 3,
    ext:       'xlsx',
  },
};

const escapeXml = s => s
  .replace(/&/g,'&amp;').replace(/</g,'&lt;')
  .replace(/>/g,'&gt;').replace(/"/g,'&quot;').replace(/'/g,'&apos;');

function fillSheet(xml, idxDesc, idxPrecio, products) {
  // Patrón exacto confirmado en ambas plantillas:
  // DESCRIPCION: <c r="B2" s="17" t="s"><v>4</v></c>
  // PRECIO:      <c r="B3" s="18" t="s"><v>5</v></c>

  const descRegex  = new RegExp(`<c r="[A-Z]+\\d+" s="\\d+" t="s"><v>${idxDesc}<\\/v><\\/c>`, 'g');
  const precioRegex = new RegExp(`<c r="[A-Z]+\\d+" s="\\d+" t="s"><v>${idxPrecio}<\\/v><\\/c>`, 'g');

  let dIdx = 0;
  let pIdx = 0;

  // Reemplazar DESCRIPCION → inline string con el nombre
  xml = xml.replace(descRegex, (match) => {
    const name = dIdx < products.length ? escapeXml(products[dIdx].name) : '';
    dIdx++;
    // Cambiar t="s" a t="str" y <v>4</v> por el nombre
    return match
      .replace(`t="s"`, `t="str"`)
      .replace(`<v>${idxDesc}</v>`, `<v>${name}</v>`);
  });

  // Reemplazar PRECIO → número (quitar t="s", poner valor numérico)
  xml = xml.replace(precioRegex, (match) => {
    if (pIdx >= products.length) {
      pIdx++;
      // Slot vacío — quitar shared string, dejar número vacío
      return match
        .replace(` t="s"`, '')
        .replace(`<v>${idxPrecio}</v>`, `<v></v>`);
    }
    const price = products[pIdx++].price;
    return match
      .replace(` t="s"`, '')
      .replace(`<v>${idxPrecio}</v>`, `<v>${price}</v>`);
  });

  return xml;
}

export default async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  if (req.method === 'OPTIONS') return res.status(200).end();
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' });

  try {
    const { cambios, template } = req.body;
    if (!cambios || !template)
      return res.status(400).json({ error: 'Faltan parámetros' });

    const cfg = TPL_CONFIG[template];
    if (!cfg) return res.status(400).json({ error: 'Plantilla desconocida: ' + template });

    // ── 1. LEER PRODUCTOS DEL ARCHIVO DE CAMBIOS ─────────────
    const cambiosBuf = Buffer.from(cambios, 'base64');
    const wbC        = XLSX.read(cambiosBuf, { type: 'buffer' });
    const sheetName  = wbC.SheetNames[wbC.SheetNames.length - 1];
    const wsC        = wbC.Sheets[sheetName];
    const rows       = XLSX.utils.sheet_to_json(wsC, { header: 1, defval: null, raw: true });

    const products = [];
    for (let i = 2; i < rows.length; i++) {
      const name = rows[i][1] != null ? String(rows[i][1]).trim().toUpperCase() : '';
      if (!name) continue;
      const price = parseFloat(String(rows[i][3] ?? '').replace(/[^\d.]/g, ''));
      if (isNaN(price)) continue;
      products.push({ name, price });
    }

    if (!products.length)
      return res.status(400).json({ error: `Sin productos en hoja "${sheetName}"` });

    // ── 2. ABRIR PLANTILLA COMO ZIP Y EDITAR XML ──────────────
    const tplPath = path.join(__dirname, '..', 'plantillas', cfg.file);
    const zip     = new AdmZip(tplPath);

    for (const sheetPath of cfg.sheets) {
      const entry = zip.getEntry(sheetPath);
      if (!entry) continue;
      const xml    = entry.getData().toString('utf8');
      const newXml = fillSheet(xml, cfg.idxDesc, cfg.idxPrecio, products);
      zip.updateFile(sheetPath, Buffer.from(newXml, 'utf8'));
    }

    // ── 3. DEVOLVER ARCHIVO MODIFICADO ────────────────────────
    const outBuf   = zip.toBuffer();
    const filename = `PRECIOS_${template.toUpperCase()}_${sheetName.replace(/\s/g, '-')}.${cfg.ext}`;

    return res.status(200).json({
      ok:       true,
      file:     outBuf.toString('base64'),
      filename,
      ext:      cfg.ext,
      count:    Math.min(products.length, 99),
      sheet:    sheetName,
    });

  } catch (err) {
    console.error('fill-precios error:', err);
    return res.status(500).json({ error: err.message });
  }
}
